fix Fatal Errors: don't assume set_ini() exists

maltfield opened the pull request:
* https://core.trac.wordpress.org/ticket/48693
 * https://core.trac.wordpress.org/ticket/62047

<!--
Hi there! Thanks for contributing to WordPress!

Pull Requests in this GitHub repository **must** be linked to a ticket in the WordPress Core Trac instance (https://core.trac.wordpress.org), and are only used for code review. **No pull requests will be merged on GitHub.**

See the WordPress Handbook page on using PRs for Code Review more information: https://make.wordpress.org/core/handbook/contribute/git/github-pull-requests-for-code-review/

If this is your first time contributing, you may also find reviewing these guides first to be helpful:
- FAQs for New Contributors: https://make.wordpress.org/core/handbook/tutorials/faq-for-new-contributors/
- Contributing with Code Guide: https://make.wordpress.org/core/handbook/contribute/
- WordPress Coding Standards: https://make.wordpress.org/core/handbook/best-practices/coding-standards/
- Inline Documentation Standards: https://make.wordpress.org/core/handbook/best-practices/inline-documentation-standards/
- Browser Support Policies: https://make.wordpress.org/core/handbook/best-practices/browser-support/
- Proper spelling and grammar related best practices: https://make.wordpress.org/core/handbook/best-practices/spelling/
-->

<!-- Insert a description of your changes here -->

Trac ticket: <!-- insert a link to the WordPress Trac ticket here -->

---
**This Pull Request is for code review only. Please keep all other discussion in the Trac ticket. Do not merge this Pull Request. See [GitHub Pull Requests for Code Review](https://make.wordpress.org/core/handbook/contribute/git/github-pull-requests-for-code-review/) in the Core Handbook for more details.**

maltfield committed:
fix Fatal Errors: don't assume set_ini() exists

 * https://core.trac.wordpress.org/ticket/62047
 * https://core.trac.wordpress.org/ticket/48693

aristath commented on src/wp-admin/plugins.php:
```diff
@@ -189,7 +189,9 @@
 			}
 
-			ini_set( 'display_errors', true ); // Ensure that fatal errors are displayed.
+			if ( function_exists('ini_set') ) {
```
```suggestion
			if ( function_exists( 'ini_set' ) ) {
```

aristath commented on src/wp-admin/update.php:
```diff
@@ -96,7 +96,9 @@
 
 			error_reporting( E_CORE_ERROR | E_CORE_WARNING | E_COMPILE_ERROR | E_ERROR | E_WARNING | E_PARSE | E_USER_ERROR | E_USER_WARNING | E_RECOVERABLE_ERROR );
-			ini_set( 'display_errors', true ); // Ensure that fatal errors are displayed.
+			if ( function_exists('ini_set') ) {
```
```suggestion
			if ( function_exists( 'ini_set' ) ) {
```

aristath commented on src/wp-includes/PHPMailer/PHPMailer.php:
I'm not sure we should make any changes to this file directly, as it's a 3rd-party library... 🤔

aristath commented on src/wp-includes/default-constants.php:
```diff
@@ -68,7 +68,9 @@ function wp_initial_constants() {
 	$wp_limit_int = wp_convert_hr_to_bytes( WP_MEMORY_LIMIT );
 	if ( -1 !== $current_limit_int && ( -1 === $wp_limit_int || $wp_limit_int > $current_limit_int ) ) {
-		ini_set( 'memory_limit', WP_MEMORY_LIMIT );
+		if ( function_exists('ini_set') ) {
```
```suggestion
		if ( function_exists( 'ini_set' ) ) {
```

aristath commented on src/wp-includes/functions.php:
```diff
@@ -7862,13 +7862,13 @@ function wp_raise_memory_limit( $context = 'admin' ) {
 
 	if ( -1 === $filtered_limit_int || ( $filtered_limit_int > $wp_max_limit_int && $filtered_limit_int > $current_limit_int ) ) {
-		if ( false !== ini_set( 'memory_limit', $filtered_limit ) ) {
+		if ( function_exists('ini_set') && false !== ini_set( 'memory_limit', $filtered_limit ) ) {
```
```suggestion
		if ( function_exists( 'ini_set' ) && false !== ini_set( 'memory_limit', $filtered_limit ) ) {
```

aristath commented on src/wp-includes/functions.php:
```diff
@@ -7862,13 +7862,13 @@ function wp_raise_memory_limit( $context = 'admin' ) {
 		}
 	} elseif ( -1 === $wp_max_limit_int || $wp_max_limit_int > $current_limit_int ) {
-		if ( false !== ini_set( 'memory_limit', $wp_max_limit ) ) {
+		if ( function_exists('ini_set') && false !== ini_set( 'memory_limit', $wp_max_limit ) ) {
```
```suggestion
		if ( function_exists( 'ini_set' ) && false !== ini_set( 'memory_limit', $wp_max_limit ) ) {
```

aristath commented on src/wp-includes/load.php:
```diff
@@ -585,9 +585,13 @@ function wp_debug_mode() {
 
 		if ( WP_DEBUG_DISPLAY ) {
-			ini_set( 'display_errors', 1 );
+			if ( function_exists('ini_set') ) {
```
```suggestion
			if ( function_exists( 'ini_set' ) ) {
```

aristath commented on src/wp-includes/load.php:
```diff
@@ -585,9 +585,13 @@ function wp_debug_mode() {
+			}
 		} elseif ( null !== WP_DEBUG_DISPLAY ) {
-			ini_set( 'display_errors', 0 );
+			if ( function_exists('ini_set') ) {
```
```suggestion
			if ( function_exists( 'ini_set' ) ) {
```

aristath commented on src/wp-includes/load.php:
```diff
@@ -599,8 +603,10 @@ function wp_debug_mode() {
 		if ( $log_path ) {
-			ini_set( 'log_errors', 1 );
-			ini_set( 'error_log', $log_path );
+			if ( function_exists('ini_set') ) {
```
```suggestion
			if ( function_exists( 'ini_set' ) ) {
```

aristath commented on src/wp-includes/load.php:
```diff
@@ -614,7 +620,9 @@ function wp_debug_mode() {
 		|| wp_doing_ajax() || wp_is_json_request()
 	) {
-		ini_set( 'display_errors', 0 );
+		if ( function_exists('ini_set') ) {
```
```suggestion
		if ( function_exists( 'ini_set' ) ) {
```

aristath commented on src/wp-includes/pomo/po.php:
```diff
@@ -21,7 +21,9 @@
  * This fix should be revisited when PHP 9.0 is in alpha/beta.
  */
-@ini_set( 'auto_detect_line_endings', 1 ); // phpcs:ignore WordPress.PHP.NoSilencedErrors.Discouraged
+if ( function_exists('ini_set') ) {
```
```suggestion
if ( function_exists( 'ini_set' ) ) {
```

maltfield committed:
Update src/wp-admin/plugins.php

Co-authored-by: Ari Stathopoulos <aristath@gmail.com>

maltfield committed:
Update src/wp-admin/update.php

Co-authored-by: Ari Stathopoulos <aristath@gmail.com>

maltfield committed:
Update src/wp-includes/default-constants.php

Co-authored-by: Ari Stathopoulos <aristath@gmail.com>

maltfield committed:
Update src/wp-includes/load.php

Co-authored-by: Ari Stathopoulos <aristath@gmail.com>

maltfield committed:
Update src/wp-includes/functions.php

Co-authored-by: Ari Stathopoulos <aristath@gmail.com>

maltfield committed:
Update src/wp-includes/pomo/po.php

Co-authored-by: Ari Stathopoulos <aristath@gmail.com>

maltfield committed:
Update src/wp-includes/load.php

Co-authored-by: Ari Stathopoulos <aristath@gmail.com>

maltfield committed:
Update src/wp-includes/load.php

Co-authored-by: Ari Stathopoulos <aristath@gmail.com>

maltfield committed:
Update src/wp-includes/load.php

Co-authored-by: Ari Stathopoulos <aristath@gmail.com>

maltfield committed:
Update src/wp-includes/functions.php

Co-authored-by: Ari Stathopoulos <aristath@gmail.com>

maltfield committed:
check to see chmod exists before using it

this commit fixes bug #62693, which prevents users from uploading media to wordpress because wordpress throws a fatal error when calling 'chmod' on servers with php.ini security-hardened following best practices, which disables dangerous functions, such as exec(), chmod(), etc.

 * https://core.trac.wordpress.org/ticket/62693

this change doesn't make the assumption that the chmod() function is available for use; it checks it first with function_exists()

maltfield committed:
Merge branch 'WordPress:trunk' into trunk

maltfield committed:
check to see chmod exists before using it

this commit fixes bug #62693, which prevents users from uploading media to wordpress because wordpress throws a fatal error when calling 'chmod' on servers with php.ini security-hardened following best practices, which disables dangerous functions, such as exec(), chmod(), etc.

 * https://core.trac.wordpress.org/ticket/62693

this change doesn't make the assumption that the chmod() function is available for use; it checks it first with function_exists()

maltfield commented:
I couldn't figure out how to tell GitHub to make a different PR for this and the new bug:

 * https://core.trac.wordpress.org/ticket/62693

So both of these bug fixes are combined into one PR.
